[BUG] setuptools version 73.0.0 breaks Pythran

### setuptools version

setuptools==73.0.0 (latest released version available on PyPI)

### Python version

All supported ones

### OS

any

### Additional environment information

This currently occurs with the latest release of Pythran (version 0.16.1 as [released on PyPI](https://pypi.org/project/pythran/0.16.1/#history) on 28/05/2024), and older versions have not been checked against.

### Description

Hi there! The latest release of `setuptools` currently breaks SciPy' builds (https://github.com/scipy/scipy/issues/21416) and, in turn, SciPy's recipe in Pyodide (https://github.com/pyodide/pyodide/issues/5018) because SciPy now uses Pythran as a required dependency for transpiling Python to C++ (as reported in https://github.com/serge-sans-paille/pythran/issues/2228).

The truncated output from Pythran is as follows – detailed output(s) below:

```
TypeError: metaclass conflict: the metaclass of a derived class must be a       
(non-strict) subclass of the metaclasses of all its bases
```

I have bisected the problem between the v72.2.0 and v73.3.0 tags as reported in https://github.com/scipy/scipy/issues/21416#issuecomment-2299197862, and the breakage seems to be coming from https://github.com/pypa/setuptools/pull/4503 (i.e., https://github.com/pypa/setuptools/commit/f9398b1f2bd6cce0aa4efa01d8b69da021205151). I have opened this issue for visibility while I continue to investigate the problem at the time of writing.

Also, please refer to @rgommers's comments in https://github.com/serge-sans-paille/pythran/issues/2228#issuecomment-2299000192 that describe the situation in context to SciPy.

### Expected behavior

The expected behaviour is that Pythran works and can be used to compile extension modules, whether for SciPy or elsewhere for all recent-enough versions of `setuptools`, since `setuptools` is a dependency of Pythran.

A workaround for this is available via https://github.com/serge-sans-paille/pythran/pull/2229 which pins `setuptools` to <73, and I have another hotfix ready at https://github.com/serge-sans-paille/pythran/pull/2230 which passes the tests.

### How to Reproduce

Here is a minimal reproduction script:

```bash
# Set up the Pythran repository
git clone https://github.com/serge-sans-paille/pythran.git pythran/
cd pythran/
# Activate a virtual environment, install the project, etc.
python3.X -m virtualenv venv
source venv/bin/activate
pip install -e ."[doc,test]"
# Run the test suite (can take long, running just one file works as well)
python -m pytest pythran/tests/test_list.py
```

### Output

```console
❯ python -m pytest pythran/tests/test_list.py      
============================================================================ test session starts ============================================================================
platform darwin -- Python 3.12.4, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/agriyakhetarpal/Desktop/pythran
configfile: pyproject.toml
plugins: nbval-0.11.0, xdist-3.6.1
collected 0 items / 1 error                                                                                                                                                 

================================================================================== ERRORS ===================================================================================
________________________________________________________________ ERROR collecting pythran/tests/test_list.py ________________________________________________________________
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1387: in _gcd_import
    ???
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1310: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:488: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1387: in _gcd_import
    ???
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:995: in exec_module
    ???
<frozen importlib._bootstrap>:488: in _call_with_frames_removed
    ???
pythran/tests/__init__.py:27: in <module>
    from pythran import compile_pythrancode, spec_parser, load_specfile, frontend
<frozen importlib._bootstrap>:1412: in _handle_fromlist
    ???
pythran/__init__.py:127: in __getattr__
    import pythran.toolchain
pythran/toolchain.py:11: in <module>
    from pythran.dist import PythranExtension, PythranBuildExt
pythran/dist.py:142: in <module>
    class PythranBuildExt(PythranBuildExtMixIn, LegacyBuildExt, metaclass=PythranBuildExtMeta):
E   TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
========================================================================== short test summary info ==========================================================================
ERROR pythran/tests/test_list.py - TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================= 1 error in 0.40s ==============================================================================
venv(base) 
pythran on  master [$] via 🐍 v3.12.4 (venv) 
❯ uv pip install "setuptools>=73"
Resolved 1 package in 16ms
Uninstalled 1 package in 0.80ms
Installed 1 package in 13ms
 - setuptools==72.2.0.post20240820 (from file:///Users/agriyakhetarpal/Desktop/setuptools)
 + setuptools==73.0.0
venv(base) 
pythran on  master [$] via 🐍 v3.12.4 (venv) 
❯ python -m pytest pythran/tests/test_list.py
============================================================================ test session starts ============================================================================
platform darwin -- Python 3.12.4, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/agriyakhetarpal/Desktop/pythran
configfile: pyproject.toml
plugins: nbval-0.11.0, xdist-3.6.1
collected 0 items / 1 error                                                                                                                                                 

================================================================================== ERRORS ===================================================================================
________________________________________________________________ ERROR collecting pythran/tests/test_list.py ________________________________________________________________
/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1387: in _gcd_import
    ???
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1310: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:488: in _call_with_frames_removed
    ???
<frozen importlib._bootstrap>:1387: in _gcd_import
    ???
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:995: in exec_module
    ???
<frozen importlib._bootstrap>:488: in _call_with_frames_removed
    ???
pythran/tests/__init__.py:27: in <module>
    from pythran import compile_pythrancode, spec_parser, load_specfile, frontend
<frozen importlib._bootstrap>:1412: in _handle_fromlist
    ???
pythran/__init__.py:127: in __getattr__
    import pythran.toolchain
pythran/toolchain.py:11: in <module>
    from pythran.dist import PythranExtension, PythranBuildExt
pythran/dist.py:142: in <module>
    class PythranBuildExt(PythranBuildExtMixIn, LegacyBuildExt, metaclass=PythranBuildExtMeta):
E   TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

--- a/newsfragments/4579.bugfix.rst
+++ b/newsfragments/4579.bugfix.rst
@@ -1,0 +1,1 @@
+Remove `abc.ABCMeta` metaclass from abstract classes. `pypa/setuptools#4503 <https://github.com/pypa/setuptools/pull/4503>`_ had an unintended consequence of causing potential ``TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`` -- by :user:`Avasam`

--- a/pkg_resources/__init__.py
+++ b/pkg_resources/__init__.py
@@ -23,7 +23,6 @@ This module is deprecated. Users are directed to :mod:`importlib.resources`,
 from __future__ import annotations
 
 import sys
-from abc import ABC
 
 if sys.version_info < (3, 8):  # noqa: UP036 # Check for unsupported versions
     raise RuntimeError("Python 3.8 or later is required")
@@ -306,7 +305,7 @@ __all__ = [
 ]
 
 
-class ResolutionError(Exception, ABC):
+class ResolutionError(Exception):
     """Abstract base for dependency resolution errors"""
 
     def __repr__(self):

--- a/setuptools/__init__.py
+++ b/setuptools/__init__.py
@@ -6,7 +6,7 @@ import functools
 import os
 import re
 import sys
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 from typing import TYPE_CHECKING, TypeVar, overload
 
 sys.path.extend(((vendor_path := os.path.join(os.path.dirname(os.path.dirname(__file__)), 'setuptools', '_vendor')) not in sys.path) * [vendor_path])  # fmt: skip
@@ -120,7 +120,7 @@ else:
     _Command = monkey.get_unpatched(distutils.core.Command)
 
 
-class Command(_Command, ABC):
+class Command(_Command):
     """
     Setuptools internal actions are organized using a *command design pattern*.
     This means that each action (or group of closely related actions) executed during

--- a/setuptools/command/setopt.py
+++ b/setuptools/command/setopt.py
@@ -1,6 +1,5 @@
 import configparser
 import os
-from abc import ABC
 
 from .. import Command
 from ..unicode_utils import _cfg_read_utf8_with_fallback
@@ -70,7 +69,7 @@ def edit_config(filename, settings, dry_run=False):
             opts.write(f)
 
 
-class option_base(Command, ABC):
+class option_base(Command):
     """Abstract base class for commands that mess with config files"""
 
     user_options = [

--- a/setuptools/sandbox.py
+++ b/setuptools/sandbox.py
@@ -11,7 +11,6 @@ import re
 import sys
 import tempfile
 import textwrap
-from abc import ABC
 
 import pkg_resources
 from pkg_resources import working_set
@@ -263,7 +262,7 @@ def run_setup(setup_script, args):
             # Normal exit, just return
 
 
-class AbstractSandbox(ABC):
+class AbstractSandbox:
     """Wrap 'os' module and 'open()' builtin for virtualizing setup scripts"""
 
     _active = False

